File: src/main/java/io/github/mcpselenium/BrowserSession.java
package io.github.mcpselenium;

import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.util.Base64;
import java.util.List;
import java.util.Locale;
import java.util.Set;

import org.openqa.selenium.By;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.Keys;
import org.openqa.selenium.OutputType;
import org.openqa.selenium.TakesScreenshot;
import org.openqa.selenium.TimeoutException;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.firefox.FirefoxOptions;
import org.openqa.selenium.interactions.Actions;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.Select;
import org.openqa.selenium.support.ui.WebDriverWait;

public final class BrowserSession {

    private static final Duration DEFAULT_TIMEOUT = Duration.ofSeconds(15);
    private static final int MAX_SCRIPT_LENGTH = 4000;

    private WebDriver driver;

    public synchronized String startBrowser(String browser, boolean headless) {
        closeBrowser();

        String selectedBrowser = browser == null
                ? "chrome"
                : browser.toLowerCase(Locale.ROOT);

        switch (selectedBrowser) {
            case "chrome": {
                ChromeOptions options = new ChromeOptions();

                if (headless) {
                    options.addArguments("--headless=new");
                }

                options.addArguments(
                        "--disable-notifications",
                        "--disable-popup-blocking",
                        "--start-maximized");

                driver = new ChromeDriver(options);
                break;
            }

            case "firefox": {
                FirefoxOptions options = new FirefoxOptions();

                if (headless) {
                    options.addArguments("-headless");
                }

                driver = new FirefoxDriver(options);
                break;
            }

            default:
                throw new IllegalArgumentException(
                        "Unsupported browser: " + selectedBrowser);
        }

        driver.manage().timeouts().pageLoadTimeout(Duration.ofSeconds(60));

        return selectedBrowser + " browser started successfully";
    }

    public synchronized String navigate(String url) {
        requireDriver();

        if (url == null || url.isBlank()) {
            throw new IllegalArgumentException("url is required");
        }

        driver.navigate().to(url);

        return "Navigated to " + url
                + ". Title: " + driver.getTitle()
                + ". Current URL: " + driver.getCurrentUrl();
    }

    public synchronized String getPageInformation() {
        requireDriver();

        return "Title: " + driver.getTitle()
                + System.lineSeparator()
                + "URL: " + driver.getCurrentUrl();
    }

    public synchronized String click(
            String strategy,
            String value,
            int timeoutSeconds) {

        WebElement element = waitForClickable(
                locator(strategy, value),
                timeoutSeconds);

        element.click();

        return "Clicked element using " + strategy + "=" + value;
    }

    public synchronized String type(
            String strategy,
            String value,
            String text,
            boolean clearFirst,
            int timeoutSeconds) {

        WebElement element = waitForVisible(
                locator(strategy, value),
                timeoutSeconds);

        if (clearFirst) {
            element.clear();
        }

        element.sendKeys(text);

        return "Entered text into element using " + strategy + "=" + value;
    }

    public synchronized String getText(
            String strategy,
            String value,
            int timeoutSeconds) {

        WebElement element = waitForVisible(
                locator(strategy, value),
                timeoutSeconds);

        return element.getText();
    }

    public synchronized String getAttribute(
            String strategy,
            String value,
            String attribute,
            int timeoutSeconds) {

        WebElement element = waitForVisible(
                locator(strategy, value),
                timeoutSeconds);

        return element.getAttribute(attribute);
    }

    public synchronized String findElement(
        String strategy,
        String value,
        int timeoutSeconds) {

        WebElement element = waitForLocatedElement(
            strategy,
            value,
            timeoutSeconds);

        return summarizeElement(element);
    }
    public synchronized String findElements(
            String strategy,
            String value,
            int timeoutSeconds) {

        By by = locator(strategy, value);

        try {
            new WebDriverWait(driver, timeout(timeoutSeconds))
                    .until(currentDriver ->
                            !currentDriver.findElements(by).isEmpty());
        } catch (TimeoutException ignored) {
            return "Found 0 elements using " + strategy + "=" + value;
        }

        List<WebElement> elements = driver.findElements(by);
        StringBuilder result = new StringBuilder();

        result.append("Found ")
                .append(elements.size())
                .append(" elements using ")
                .append(strategy)
                .append("=")
                .append(value);

        for (int index = 0; index < elements.size(); index++) {
            WebElement element = elements.get(index);
            result.append(System.lineSeparator())
                    .append(index)
                    .append(": tag=")
                    .append(element.getTagName())
                    .append(", displayed=")
                    .append(element.isDisplayed())
                    .append(", enabled=")
                    .append(element.isEnabled())
                    .append(", text=")
                    .append(trim(element.getText(), 120));
        }

        return result.toString();
    }

    public synchronized String elementExists(
            String strategy,
            String value,
            int timeoutSeconds) {

        try {
            waitForVisible(locator(strategy, value), timeoutSeconds);
            return "true";
        } catch (TimeoutException exception) {
            return "false";
        }
    }

    public synchronized String waitForElement(
            String strategy,
            String value,
            String state,
            int timeoutSeconds) {

        By by = locator(strategy, value);
        String expectedState = state == null || state.isBlank()
                ? "visible"
                : state.toLowerCase(Locale.ROOT);

        switch (expectedState) {
            case "present":
                new WebDriverWait(driver, timeout(timeoutSeconds))
                        .until(ExpectedConditions.presenceOfElementLocated(by));
                break;
            case "visible":
                waitForVisible(by, timeoutSeconds);
                break;
            case "clickable":
                waitForClickable(by, timeoutSeconds);
                break;
            default:
                throw new IllegalArgumentException(
                        "Unsupported element state: " + state);
        }

        return "Element is " + expectedState
                + " using " + strategy + "=" + value;
    }

    public synchronized String selectOption(
            String strategy,
            String value,
            String selectBy,
            String option,
            int timeoutSeconds) {

        Select select = new Select(waitForVisible(
                locator(strategy, value),
                timeoutSeconds));

        String method = selectBy.toLowerCase(Locale.ROOT);

        switch (method) {
            case "text":
                select.selectByVisibleText(option);
                break;
            case "value":
                select.selectByValue(option);
                break;
            case "index":
                select.selectByIndex(Integer.parseInt(option));
                break;
            default:
                throw new IllegalArgumentException(
                        "Unsupported selectBy value: " + selectBy);
        }

        return "Selected option by " + selectBy
                + " from element using " + strategy + "=" + value;
    }

    public synchronized String switchToFrame(
            String strategy,
            String value,
            String nameOrId,
            Integer index,
            int timeoutSeconds) {

        requireDriver();

        if (index != null) {
            driver.switchTo().frame(index);
            return "Switched to frame index " + index;
        }

        if (nameOrId != null && !nameOrId.isBlank()) {
            driver.switchTo().frame(nameOrId);
            return "Switched to frame " + nameOrId;
        }

        WebElement frame = waitForVisible(
                locator(strategy, value),
                timeoutSeconds);

        driver.switchTo().frame(frame);

        return "Switched to frame using " + strategy + "=" + value;
    }

    public synchronized String switchToDefaultContent() {
        requireDriver();
        driver.switchTo().defaultContent();
        return "Switched to default content";
    }

    public synchronized String getWindowHandles() {
        requireDriver();

        String currentHandle = driver.getWindowHandle();
        Set<String> handles = driver.getWindowHandles();

        return "Current window: " + currentHandle
                + System.lineSeparator()
                + "Window handles: " + handles;
    }

    public synchronized String switchToWindow(String handle) {
        requireDriver();

        if (handle == null || handle.isBlank()) {
            throw new IllegalArgumentException("handle is required");
        }

        driver.switchTo().window(handle);

        return "Switched to window " + handle
                + ". Title: " + driver.getTitle()
                + ". URL: " + driver.getCurrentUrl();
    }

    public synchronized String executeScript(
            String script,
            String strategy,
            String value,
            int timeoutSeconds) {

        requireDriver();

        if (script == null || script.isBlank()) {
            throw new IllegalArgumentException("script is required");
        }

        if (script.length() > MAX_SCRIPT_LENGTH) {
            throw new IllegalArgumentException(
                    "script must be " + MAX_SCRIPT_LENGTH
                            + " characters or fewer");
        }

        if (!(driver instanceof JavascriptExecutor)) {
            throw new IllegalStateException(
                    "Current driver does not support JavaScript execution");
        }

        JavascriptExecutor executor = (JavascriptExecutor) driver;

        Object result;

        if (strategy != null && !strategy.isBlank()) {
            WebElement element = waitForVisible(
                    locator(strategy, value),
                    timeoutSeconds);
            result = executor.executeScript(script, element);
        } else {
            result = executor.executeScript(script);
        }

        return result == null ? "null" : result.toString();
    }

    public synchronized String scrollToElement(
            String strategy,
            String value,
            int timeoutSeconds) {

        WebElement element = waitForVisible(
                locator(strategy, value),
                timeoutSeconds);

        javascriptExecutor().executeScript(
                "arguments[0].scrollIntoView({block:'center',inline:'nearest'});",
                element);

        return "Scrolled to element using " + strategy + "=" + value;
    }

    public synchronized String hoverElement(
            String strategy,
            String value,
            int timeoutSeconds) {

        WebElement element = waitForVisible(
                locator(strategy, value),
                timeoutSeconds);

        new Actions(driver).moveToElement(element).perform();

        return "Hovered over element using " + strategy + "=" + value;
    }

    public synchronized String pressKey(
            String key,
            String strategy,
            String value,
            int timeoutSeconds) {

        requireDriver();

        Keys seleniumKey = keyValue(key);

        if (strategy != null && !strategy.isBlank()) {
            WebElement element = waitForVisible(
                    locator(strategy, value),
                    timeoutSeconds);
            element.sendKeys(seleniumKey);
        } else {
            new Actions(driver).sendKeys(seleniumKey).perform();
        }

        return "Pressed key " + key;
    }

    public synchronized String uploadFile(
            String strategy,
            String value,
            String filePath,
            int timeoutSeconds) {

        if (filePath == null || filePath.isBlank()) {
            throw new IllegalArgumentException("filePath is required");
        }

        Path path = Path.of(filePath).toAbsolutePath();

        if (!Files.isRegularFile(path)) {
            throw new IllegalArgumentException(
                    "File does not exist: " + path);
        }

        WebElement element = waitForVisible(
                locator(strategy, value),
                timeoutSeconds);

        element.sendKeys(path.toString());

        return "Uploaded file " + path;
    }

    public synchronized String takeScreenshotBase64() {
        requireDriver();

        if (!(driver instanceof TakesScreenshot)) {
            throw new IllegalStateException(
                    "Current driver does not support screenshots");
        }

        TakesScreenshot screenshotDriver = (TakesScreenshot) driver;
        byte[] bytes = screenshotDriver.getScreenshotAs(OutputType.BYTES);

        return Base64.getEncoder().encodeToString(bytes);
    }

    public synchronized String getAccessibilitySnapshot() {
        requireDriver();

        String script =
            "return JSON.stringify((() => {\n"
            + "  const visible = element => {\n"
            + "    const style = window.getComputedStyle(element);\n"
            + "    const rect = element.getBoundingClientRect();\n"
            + "    return style.visibility !== 'hidden'\n"
            + "      && style.display !== 'none'\n"
            + "      && rect.width > 0\n"
            + "      && rect.height > 0;\n"
            + "  };\n\n"
            + "  const nameOf = element =>\n"
            + "    element.getAttribute('aria-label')\n"
            + "    || element.getAttribute('alt')\n"
            + "    || element.getAttribute('title')\n"
            + "    || element.innerText\n"
            + "    || element.value\n"
            + "    || '';\n\n"
            + "  const roleOf = element =>\n"
            + "    element.getAttribute('role')\n"
            + "    || ({ A: 'link', BUTTON: 'button', INPUT: element.type || 'input',\n"
            + "         SELECT: 'select', TEXTAREA: 'textbox', IMG: 'image' })[element.tagName]\n"
            + "    || '';\n\n"
            + "  const selector = 'a,button,input,select,textarea,img,[role],'\n"
            + "    + '[aria-label],[aria-labelledby],h1,h2,h3,h4,h5,h6';\n\n"
            + "  const nodes = Array.from(document.body.querySelectorAll(selector))\n"
            + "    .filter(visible).slice(0, 200)\n"
            + "    .map((element, index) => ({\n"
            + "      index, tag: element.tagName.toLowerCase(), role: roleOf(element),\n"
            + "      name: nameOf(element).replace(/\\s+/g, ' ').trim().slice(0, 160),\n"
            + "      disabled: element.disabled === true, href: element.href || null\n"
            + "    }));\n\n"
            + "  return { title: document.title, url: location.href, count: nodes.length, nodes };\n"
            + "})(), null, 2);\n";

        Object result = javascriptExecutor().executeScript(script);

        return result == null ? "{}" : result.toString();
    }

    public synchronized String getPageSource() {
        requireDriver();
        return driver.getPageSource();
    }

    public synchronized byte[] takeScreenshot() {
        requireDriver();

        if (!(driver instanceof TakesScreenshot)) {
            throw new IllegalStateException(
                    "Current driver does not support screenshots");
        }

        TakesScreenshot screenshotDriver = (TakesScreenshot) driver;
        return screenshotDriver.getScreenshotAs(OutputType.BYTES);
    }

    public synchronized String closeBrowser() {
        if (driver != null) {
            try {
                driver.quit();
            } finally {
                driver = null;
            }

            return "Browser closed";
        }

        return "No browser session was active";
    }

    private WebElement waitForLocatedElement(
        String strategy,
        String value,
        int timeoutSeconds) {

        requireDriver();

        return new WebDriverWait(driver, timeout(timeoutSeconds))
            .until(ExpectedConditions.presenceOfElementLocated(
                    locator(strategy, value)));
    }

    private String summarizeElement(WebElement element) {
        return "Element: tag="
            + element.getTagName()
            + ", displayed="
            + element.isDisplayed()
            + ", enabled="
            + element.isEnabled()
            + ", text="
            + trim(element.getText(), 120);
    }

    private WebElement waitForVisible(By by, int timeoutSeconds) {
        requireDriver();

        return new WebDriverWait(driver, timeout(timeoutSeconds))
                .until(ExpectedConditions.visibilityOfElementLocated(by));
    }

    private WebElement waitForClickable(By by, int timeoutSeconds) {
        requireDriver();

        return new WebDriverWait(driver, timeout(timeoutSeconds))
                .until(ExpectedConditions.elementToBeClickable(by));
    }

    private JavascriptExecutor javascriptExecutor() {
        requireDriver();

        if (!(driver instanceof JavascriptExecutor)) {
            throw new IllegalStateException(
                    "Current driver does not support JavaScript execution");
        }

        return (JavascriptExecutor) driver;
    }

    private Duration timeout(int timeoutSeconds) {
        return timeoutSeconds > 0
                ? Duration.ofSeconds(timeoutSeconds)
                : DEFAULT_TIMEOUT;
    }

    private By locator(String strategy, String value) {
        if (strategy == null || strategy.isBlank()) {
            throw new IllegalArgumentException(
                    "locator strategy is required");
        }

        if (value == null || value.isBlank()) {
            throw new IllegalArgumentException(
                    "locator value is required");
        }

        switch (strategy.toLowerCase(Locale.ROOT)) {
            case "id":
                return By.id(value);
            case "name":
                return By.name(value);
            case "css":
            case "cssselector":
                return By.cssSelector(value);
            case "xpath":
                return By.xpath(value);
            case "class":
            case "classname":
                return By.className(value);
            case "tag":
            case "tagname":
                return By.tagName(value);
            case "linktext":
                return By.linkText(value);
            case "partiallinktext":
                return By.partialLinkText(value);
            default:
                throw new IllegalArgumentException(
                        "Unsupported locator strategy: " + strategy);
        }
    }

    private Keys keyValue(String key) {
        if (key == null || key.isBlank()) {
            throw new IllegalArgumentException("key is required");
        }

        switch (key.toUpperCase(Locale.ROOT)) {
            case "ENTER":
                return Keys.ENTER;
            case "TAB":
                return Keys.TAB;
            case "ESCAPE":
                return Keys.ESCAPE;
            case "SPACE":
                return Keys.SPACE;
            case "BACKSPACE":
                return Keys.BACK_SPACE;
            case "DELETE":
                return Keys.DELETE;
            case "ARROW_UP":
                return Keys.ARROW_UP;
            case "ARROW_DOWN":
                return Keys.ARROW_DOWN;
            case "ARROW_LEFT":
                return Keys.ARROW_LEFT;
            case "ARROW_RIGHT":
                return Keys.ARROW_RIGHT;
            case "HOME":
                return Keys.HOME;
            case "END":
                return Keys.END;
            case "PAGE_UP":
                return Keys.PAGE_UP;
            case "PAGE_DOWN":
                return Keys.PAGE_DOWN;
            default:
                throw new IllegalArgumentException(
                        "Unsupported key: " + key);
        }
    }

    private String trim(String value, int maxLength) {
        if (value == null) {
            return "";
        }

        String normalized = value.replaceAll("\\s+", " ").trim();

        return normalized.length() <= maxLength
                ? normalized
                : normalized.substring(0, maxLength) + "...";
    }

    private void requireDriver() {
        if (driver == null) {
            throw new IllegalStateException(
                    "No browser session. Call start_browser first.");
        }
    }
}

File: python/browser_session.py
"""Reusable Selenium automation engine (Python port of BrowserSession.java).

This module owns all direct WebDriver interaction: session lifecycle, locators,
waits, element actions, and page inspection. It has no knowledge of MCP or of
any agent orchestration layer, so it can be reused by any tool/API surface.
"""

from __future__ import annotations

import base64
import json
import threading
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

from selenium import webdriver
from selenium.common.exceptions import TimeoutException
from selenium.webdriver.common.action_chains import ActionChains
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.support.select import Select
from selenium.webdriver.support.ui import WebDriverWait

DEFAULT_TIMEOUT_SECONDS = 15
MAX_SCRIPT_LENGTH = 4000

_LOCATOR_STRATEGIES = {
    "id": By.ID,
    "name": By.NAME,
    "css": By.CSS_SELECTOR,
    "cssselector": By.CSS_SELECTOR,
    "xpath": By.XPATH,
    "class": By.CLASS_NAME,
    "classname": By.CLASS_NAME,
    "tag": By.TAG_NAME,
    "tagname": By.TAG_NAME,
    "linktext": By.LINK_TEXT,
    "partiallinktext": By.PARTIAL_LINK_TEXT,
}

_KEYS = {
    "ENTER": Keys.ENTER,
    "TAB": Keys.TAB,
    "ESCAPE": Keys.ESCAPE,
    "SPACE": Keys.SPACE,
    "BACKSPACE": Keys.BACK_SPACE,
    "DELETE": Keys.DELETE,
    "ARROW_UP": Keys.ARROW_UP,
    "ARROW_DOWN": Keys.ARROW_DOWN,
    "ARROW_LEFT": Keys.ARROW_LEFT,
    "ARROW_RIGHT": Keys.ARROW_RIGHT,
    "HOME": Keys.HOME,
    "END": Keys.END,
    "PAGE_UP": Keys.PAGE_UP,
    "PAGE_DOWN": Keys.PAGE_DOWN,
}

_ACCESSIBILITY_SNAPSHOT_SCRIPT = """
return JSON.stringify((() => {
  const visible = element => {
    const style = window.getComputedStyle(element);
    const rect = element.getBoundingClientRect();
    return style.visibility !== 'hidden'
      && style.display !== 'none'
      && rect.width > 0
      && rect.height > 0;
  };
  const nameOf = element =>
    element.getAttribute('aria-label')
    || element.getAttribute('alt')
    || element.getAttribute('title')
    || element.innerText
    || element.value
    || '';
  const roleOf = element =>
    element.getAttribute('role')
    || ({
      A: 'link', BUTTON: 'button', INPUT: element.type || 'input',
      SELECT: 'select', TEXTAREA: 'textbox', IMG: 'image'
    })[element.tagName]
    || '';
  const selector =
    'a,button,input,select,textarea,img,[role],'
    + '[aria-label],[aria-labelledby],h1,h2,h3,h4,h5,h6';
  const nodes = Array.from(document.body.querySelectorAll(selector))
    .filter(visible)
    .slice(0, 200)
    .map((element, index) => ({
      index,
      tag: element.tagName.toLowerCase(),
      role: roleOf(element),
      name: nameOf(element).replace(/\\s+/g, ' ').trim().slice(0, 160),
      disabled: element.disabled === true,
      href: element.href || null
    }));
  return { title: document.title, url: location.href, count: nodes.length, nodes };
})(), null, 2);
"""


def _trim(text: Optional[str], limit: int) -> str:
    if not text:
        return ""
    return text if len(text) <= limit else text[:limit] + "..."


@dataclass
class BrowserSession:
    """Owns a single WebDriver instance. Not thread-safe across concurrent
    sessions by design -- one session per agent/user, guarded by a lock."""

    driver: Optional[webdriver.Remote] = None

    def __post_init__(self) -> None:
        self._lock = threading.RLock()

    # -- lifecycle ---------------------------------------------------------

    def start_browser(self, browser: str, headless: bool = False) -> str:
        with self._lock:
            self.close_browser()

            selected = (browser or "chrome").lower()

            if selected == "chrome":
                options = webdriver.ChromeOptions()
                if headless:
                    options.add_argument("--headless=new")
                options.add_argument("--disable-notifications")
                options.add_argument("--disable-popup-blocking")
                options.add_argument("--start-maximized")
                self.driver = webdriver.Chrome(options=options)
            elif selected == "firefox":
                options = webdriver.FirefoxOptions()
                if headless:
                    options.add_argument("-headless")
                self.driver = webdriver.Firefox(options=options)
            else:
                raise ValueError(f"Unsupported browser: {selected}")

            self.driver.set_page_load_timeout(60)
            return f"{selected} browser started successfully"

    def close_browser(self) -> str:
        with self._lock:
            if self.driver is not None:
                try:
                    self.driver.quit()
                finally:
                    self.driver = None
                return "Browser closed"
            return "No browser session was active"

    # -- navigation ----------------------------------------------------------

    def navigate(self, url: str) -> str:
        driver = self._require_driver()
        if not url:
            raise ValueError("url is required")
        driver.get(url)
        return f"Navigated to {url}. Title: {driver.title}. Current URL: {driver.current_url}"

    def get_page_information(self) -> str:
        driver = self._require_driver()
        return f"Title: {driver.title}\nURL: {driver.current_url}"

    # -- element actions -----------------------------------------------------

    def click(self, strategy: str, value: str, timeout_seconds: int = DEFAULT_TIMEOUT_SECONDS) -> str:
        element = self._wait_for_clickable(self._locator(strategy, value), timeout_seconds)
        element.click()
        return f"Clicked element using {strategy}={value}"

    def type_text(
        self,
        strategy: str,
        value: str,
        text: str,
        clear_first: bool = True,
        timeout_seconds: int = DEFAULT_TIMEOUT_SECONDS,
    ) -> str:
        element = self._wait_for_visible(self._locator(strategy, value), timeout_seconds)
        if clear_first:
            element.clear()
        element.send_keys(text)
        return f"Entered text into element using {strategy}={value}"

    def get_text(self, strategy: str, value: str, timeout_seconds: int = DEFAULT_TIMEOUT_SECONDS) -> str:
        element = self._wait_for_visible(self._locator(strategy, value), timeout_seconds)
        return element.text

    def get_attribute(
        self, strategy: str, value: str, attribute: str, timeout_seconds: int = DEFAULT_TIMEOUT_SECONDS
    ) -> str:
        element = self._wait_for_visible(self._locator(strategy, value), timeout_seconds)
        return element.get_attribute(attribute) or ""

    def find_element(self, strategy: str, value: str, timeout_seconds: int = DEFAULT_TIMEOUT_SECONDS) -> str:
        driver = self._require_driver()
        by = self._locator(strategy, value)
        element = WebDriverWait(driver, timeout_seconds or DEFAULT_TIMEOUT_SECONDS).until(
            EC.presence_of_element_located(by)
        )
        return self._summarize_element(element)

    def find_elements(self, strategy: str, value: str, timeout_seconds: int = DEFAULT_TIMEOUT_SECONDS) -> str:
        driver = self._require_driver()
        by = self._locator(strategy, value)
        try:
            WebDriverWait(driver, timeout_seconds or DEFAULT_TIMEOUT_SECONDS).until(
                lambda d: len(d.find_elements(*by)) > 0
            )
        except TimeoutException:
            return f"Found 0 elements using {strategy}={value}"

        elements = driver.find_elements(*by)
        lines = [f"Found {len(elements)} elements using {strategy}={value}"]
        for index, element in enumerate(elements):
            lines.append(
                f"{index}: tag={element.tag_name}, displayed={element.is_displayed()}, "
                f"enabled={element.is_enabled()}, text={_trim(element.text, 120)}"
            )
        return "\n".join(lines)

    def element_exists(self, strategy: str, value: str, timeout_seconds: int = 5) -> str:
        try:
            self._wait_for_visible(self._locator(strategy, value), timeout_seconds)
            return "true"
        except TimeoutException:
            return "false"

    def wait_for_element(
        self,
        strategy: str,
        value: str,
        state: str = "visible",
        timeout_seconds: int = DEFAULT_TIMEOUT_SECONDS,
    ) -> str:
        driver = self._require_driver()
        by = self._locator(strategy, value)
        expected_state = (state or "visible").lower()

        if expected_state == "present":
            WebDriverWait(driver, timeout_seconds or DEFAULT_TIMEOUT_SECONDS).until(
                EC.presence_of_element_located(by)
            )
        elif expected_state == "visible":
            self._wait_for_visible(by, timeout_seconds)
        elif expected_state == "clickable":
            self._wait_for_clickable(by, timeout_seconds)
        else:
            raise ValueError(f"Unsupported element state: {state}")

        return f"Element is {expected_state} using {strategy}={value}"

    def select_option(
        self,
        strategy: str,
        value: str,
        select_by: str,
        option: str,
        timeout_seconds: int = DEFAULT_TIMEOUT_SECONDS,
    ) -> str:
        element = self._wait_for_visible(self._locator(strategy, value), timeout_seconds)
        select = Select(element)
        method = select_by.lower()

        if method == "text":
            select.select_by_visible_text(option)
        elif method == "value":
            select.select_by_value(option)
        elif method == "index":
            select.select_by_index(int(option))
        else:
            raise ValueError(f"Unsupported selectBy value: {select_by}")

        return f"Selected option by {select_by} from element using {strategy}={value}"

    # -- frames / windows ------------------------------------------------------

    def switch_to_frame(
        self,
        strategy: Optional[str],
        value: Optional[str],
        name_or_id: Optional[str],
        index: Optional[int],
        timeout_seconds: int = DEFAULT_TIMEOUT_SECONDS,
    ) -> str:
        driver = self._require_driver()

        if index is not None:
            driver.switch_to.frame(index)
            return f"Switched to frame index {index}"

        if name_or_id:
            driver.switch_to.frame(name_or_id)
            return f"Switched to frame {name_or_id}"

        frame = self._wait_for_visible(self._locator(strategy, value), timeout_seconds)
        driver.switch_to.frame(frame)
        return f"Switched to frame using {strategy}={value}"

    def switch_to_default_content(self) -> str:
        driver = self._require_driver()
        driver.switch_to.default_content()
        return "Switched to default content"

    def get_window_handles(self) -> str:
        driver = self._require_driver()
        return f"Current window: {driver.current_window_handle}\nWindow handles: {driver.window_handles}"

    def switch_to_window(self, handle: str) -> str:
        driver = self._require_driver()
        if not handle:
            raise ValueError("handle is required")
        driver.switch_to.window(handle)
        return f"Switched to window {handle}. Title: {driver.title}. URL: {driver.current_url}"

    # -- javascript / misc ------------------------------------------------------

    def execute_script(
        self,
        script: str,
        strategy: Optional[str] = None,
        value: Optional[str] = None,
        timeout_seconds: int = DEFAULT_TIMEOUT_SECONDS,
    ) -> str:
        driver = self._require_driver()
        if not script:
            raise ValueError("script is required")
        if len(script) > MAX_SCRIPT_LENGTH:
            raise ValueError(f"script must be {MAX_SCRIPT_LENGTH} characters or fewer")

        if strategy:
            element = self._wait_for_visible(self._locator(strategy, value), timeout_seconds)
            result = driver.execute_script(script, element)
        else:
            result = driver.execute_script(script)

        return "null" if result is None else str(result)

    def scroll_to_element(self, strategy: str, value: str, timeout_seconds: int = DEFAULT_TIMEOUT_SECONDS) -> str:
        driver = self._require_driver()
        element = self._wait_for_visible(self._locator(strategy, value), timeout_seconds)
        driver.execute_script("arguments[0].scrollIntoView({block:'center',inline:'nearest'});", element)
        return f"Scrolled to element using {strategy}={value}"

    def hover_element(self, strategy: str, value: str, timeout_seconds: int = DEFAULT_TIMEOUT_SECONDS) -> str:
        driver = self._require_driver()
        element = self._wait_for_visible(self._locator(strategy, value), timeout_seconds)
        ActionChains(driver).move_to_element(element).perform()
        return f"Hovered over element using {strategy}={value}"

    def press_key(
        self,
        key: str,
        strategy: Optional[str] = None,
        value: Optional[str] = None,
        timeout_seconds: int = DEFAULT_TIMEOUT_SECONDS,
    ) -> str:
        driver = self._require_driver()
        selenium_key = self._key_value(key)

        if strategy:
            element = self._wait_for_visible(self._locator(strategy, value), timeout_seconds)
            element.send_keys(selenium_key)
        else:
            ActionChains(driver).send_keys(selenium_key).perform()

        return f"Pressed key {key}"

    def upload_file(
        self, strategy: str, value: str, file_path: str, timeout_seconds: int = DEFAULT_TIMEOUT_SECONDS
    ) -> str:
        if not file_path:
            raise ValueError("filePath is required")

        path = Path(file_path).resolve()
        if not path.is_file():
            raise ValueError(f"File does not exist: {path}")

        element = self._wait_for_visible(self._locator(strategy, value), timeout_seconds)
        element.send_keys(str(path))
        return f"Uploaded file {path}"

    def take_screenshot_base64(self) -> str:
        driver = self._require_driver()
        return base64.b64encode(driver.get_screenshot_as_png()).decode("ascii")

    def get_accessibility_snapshot(self) -> str:
        driver = self._require_driver()
        result = driver.execute_script(_ACCESSIBILITY_SNAPSHOT_SCRIPT)
        return "{}" if result is None else str(result)

    def get_page_source(self) -> str:
        return self._require_driver().page_source

    # -- internal helpers ------------------------------------------------------

    def _require_driver(self) -> webdriver.Remote:
        if self.driver is None:
            raise RuntimeError("No active browser session. Call start_browser first.")
        return self.driver

    def _summarize_element(self, element) -> str:
        return (
            f"Element: tag={element.tag_name}, displayed={element.is_displayed()}, "
            f"enabled={element.is_enabled()}, text={_trim(element.text, 120)}"
        )

    def _wait_for_visible(self, by, timeout_seconds: int):
        driver = self._require_driver()
        return WebDriverWait(driver, timeout_seconds or DEFAULT_TIMEOUT_SECONDS).until(
            EC.visibility_of_element_located(by)
        )

    def _wait_for_clickable(self, by, timeout_seconds: int):
        driver = self._require_driver()
        return WebDriverWait(driver, timeout_seconds or DEFAULT_TIMEOUT_SECONDS).until(
            EC.element_to_be_clickable(by)
        )

    def _locator(self, strategy: Optional[str], value: Optional[str]):
        if not strategy:
            raise ValueError("locator strategy is required")
        if not value:
            raise ValueError("locator value is required")

        by = _LOCATOR_STRATEGIES.get(strategy.lower())
        if by is None:
            raise ValueError(f"Unsupported locator strategy: {strategy}")
        return (by, value)

    def _key_value(self, key: Optional[str]) -> str:
        if not key:
            raise ValueError("key is required")
        selenium_key = _KEYS.get(key.upper())
        if selenium_key is None:
            raise ValueError(f"Unsupported key: {key}")
        return selenium_key
